### ISSUE TITLE
The usage of custom widgets with @interact is unnecessarily restricted and incorrectly documented

### ISSUE THREAD

**Issue by bermani** [jupyter-widgets/ipywidgets#2511]
This issue is directly related to #2261 but I thought I would make a new issue to show all of my findings. Essentially, interact has unnecessary restrictions on the widgets that can be used, these restrictions are conflicting with the behavior specified in the documentation.

## Current Documentation

According to the [docs on interact](https://ipywidgets.readthedocs.io/en/stable/examples/Using%20Interact.html#Widget-abbreviations), 

> This examples clarifies how interact processes its keyword arguments:
>
>1. If the keyword argument is a Widget instance with a value attribute, that widget is used. Any widget with a value attribute can be used, even custom ones.
>
>2. Otherwise, the value is treated as a widget abbreviation that is converted to a widget before it is used.

We should be able to pass any Widget instance with a value attribute, however, this is not the case.

### 1. Interact errors on any widget that's not a ValueWidget
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L297-L302

The call to `widget_from_abbrev` on line 297 is intended to convert any widgets to ValueWidgets; the docstring for `widget_from_abbrev` states that it should do this:
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L310-L311

However, when examining this function and the functions it calls, there is no section that attempts to build a ValueWidget instance from a passed widget, it only either accepts a ValueWidget instance or tries to make one from abbreviations.

https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L342-L343
The function returns none, and thus the ValueError on line 300 is raised. This is a confusing and incorrect error message when passing a Widget that isn't a ValueWidget, to see "YourWidget cannot be transformed into a widget". 

### 2. Interact errors on any widget that doesn't have a description attribute
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L303-L304
I think the intention behind this section is to update the description attribute of a Widget if no description is provided, but the current implementation requires the user to put at the very least `self.description = False`. This is easy to work around if you're making your own widget but can be annoying if you want to use widgets from a library and the developer didn't put a description attribute. 

Basically, the docs say we can use any Widget with a value attribute, but the implementation requires ValueWidgets with description attributes.

## My opinion on correct behavior
I think that the right thing to do is to implement the routine that transforms Widgets into ValueWidgets, and tweak the description checking to work without description predefined. Alternatively, the docs could be just edited to reflect how the implementation currently works.

**Comment by SylvainCorlay** [comment]
Thank you @bermani for the detailed report. I agree with you that the documentation should be updated.

Although maybe it would be cleaner to actually document that a `ValueWidget` is expected. This may be more in line with other jupyter widgets backend with less dynamic languages.

**Comment by bermani** [comment]
Yes, that is a good option. I certainly think that in the back end, all widgets used with interact should be represented as ValueWidgets. I feel like we don't have to require the frontend user to pass a ValueWidget though, I thought the behavior specified in the docs and code comments was a good idea: we can convert any widget with a value attribute to a ValueWidget, thus allowing any Widget with a value attribute to be passed. This would make widgets easier to use and more accessible. I can see the argument for making the Python backend more in line with other backends, however, I am familiar with neither other backends nor the requirements for how similar they must be. 

Since this is your project, you have the decision on this matter. I'm interested in making the changes and creating a PR. @SylvainCorlay

**Comment by pbugnion** [comment]
Closed by #2544, which enforces using a ValueWidget and updates the documentation. Thank you!
